### PR TITLE
Enrich dimension targets in click_behavior with stage-numbers

### LINKED ISSUE #49110: Migrate existing click behavior to always include explicit target stage-number
Similar to #48734 but for click behavior in viz settings.

### PR DESCRIPTION
**Pull request by metamben**
Fixes #49110

### Description

Like #49256, but adding stage-numbers in dimension targets in click behavior settings.

### TODO

- [x] Make stage number depend on the target of the click, not on the dashcard's card.
- [x] Add path `["visualization_settings" "click_behavior" "parameterMapping" ["target" "dimension"]]` to the migration.
- [x] Only migrate if `click_behavior.linkType = "question"`.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by piranha** [comment]
Do you think it needs to be a migration rather than an augment during read?.. I mean data migrations are finicky enough to potentially be a problem, and we already have `migrate-viz-settings` for stuff like that. I see that there is another migration merged in, but it's also not released yet, so it's more or less in the same condition. 😁

File: src/metabase/db/custom_migrations.clj [hunk 1]
```diff
@@ -1683,6 +1683,17 @@
       (when (not= pmappings orig-pmappings)
         pmappings))))
 
+(defn- card->last-stage
```
Comment (kamilmielnik) [inline src/metabase/db/custom_migrations.clj:1686]: :x:  `stage-number` is `0` for multi-stage queries

1. `git checkout master`
2. `yarn dev`
3. https://github.com/user-attachments/assets/973ffbce-3d9a-4eee-9a00-22d7db075026
4. kill `yarn dev`
5. `git checkout 49110-click-behavior-stage-number-migration`
6. `yarn dev`
7. https://github.com/user-attachments/assets/bf937294-f280-4d5f-bc2c-5893e02b889f

**Review by kamilmielnik — CHANGES_REQUESTED** [review]

**Comment by metamben** [comment]
> Do you think it needs to be a migration rather than an augment during read?.. I mean data migrations are finicky enough to potentially be a problem, and we already have `migrate-viz-settings` for stuff like that. I see that there is another migration merged in, but it's also not released yet, so it's more or less in the same condition. 😁

Yes, we had an extended discussion about this and the outcome was that not doing this would be more error prone.

File: src/metabase/db/custom_migrations.clj (hunk 1 above)
Comment (kamilmielnik) [reply src/metabase/db/custom_migrations.clj:1686]: This is fixed now

File: src/metabase/db/custom_migrations.clj [hunk 2]
```diff
@@ -1683,6 +1683,17 @@
...
+  [record]
```
Comment (kamilmielnik) [inline src/metabase/db/custom_migrations.clj:1687]: :x: Parameter mapping disappears in the UI after running the migration.
Reason unclear.
It may have something to do with object keys not matching `target.dimension` - we should regenerate those.

image

**Review by kamilmielnik — CHANGES_REQUESTED** [review]

File: src/metabase/db/custom_migrations.clj (hunk 2 above)
Comment (metamben) [reply src/metabase/db/custom_migrations.clj:1687]: Added conversion for the `id` fields and keys.

Comment (kamilmielnik) [reply src/metabase/db/custom_migrations.clj:1687]: This is fixed now

**Review by kamilmielnik — APPROVED** [review]
I did some testing and didn’t encounter any issues :+1:
I haven’t reviewed the code though.

File: src/metabase/db/custom_migrations.clj [hunk 3]
```diff
@@ -1743,3 +1747,131 @@
...
+                         (map #(update-parameter-mapping-target-dimension % update-fn))
+                         mappings))))))
+
+(defn- update-viz-settings-target-dimensions
+  [viz-settings click-behavior->update-fn]
+  (cond-> viz-settings
+    (-> viz-settings :column_settings map?)
+    (update :column_settings
+            update-vals
+            #(update-column-settings-target-dimensions % click-behavior->update-fn))))
```
Comment (bshepherdson) [inline src/metabase/db/custom_migrations.clj:1795]: Can this be `m/update-existing` instead?

File: test/metabase/db/custom_migrations_test.clj [hunk 4]
```diff
@@ -2346,3 +2369,143 @@
...
+                    multi-stage-question-id
+                    multi-stage-model-id]}
+            (insert-stage-number-data!)
+
+            viz-settings-generator
+            (fn [dimensions]             ; 4 dimensions are consumed
+              ;; crazy as it sounds, the FE uses the JSON encoded form of the dimensions as IDs and keys
+              (let [dimension-strs (mapv json/generate-string dimensions)
+                    dimension-keys (mapv keyword dimension-strs)]
+                (fn [[target-id0 target-id1 target-id2]]
```
Comment (bshepherdson) [inline test/metabase/db/custom_migrations_test.clj:2388]: This inner function is really hard to follow, mostly because of the vertical length. Can we either abstract an inner chunk into a separate function (perhaps from index to `:parameterMapping` map?), turn it into a `for` or similar, or use the horizontal space a bit more effectively?

**Review by bshepherdson — APPROVED** [review]

File: src/metabase/db/custom_migrations.clj (hunk 3 above)
Comment (metamben) [reply src/metabase/db/custom_migrations.clj:1703]: I prefer making sure here that the value is indeed a map. I don't want to mess with anything else.
